### ISSUE TITLE
Discrepancy with es modules / duplicated multiple entries

### ISSUE THREAD

**Issue by dy** [parcel-bundler/parcel#3575]
# 🐛 bug report

Take `index.html`:
```html
<script src="./x.js"></script>
<script src="./x.js"></script>
```

```js
// x.js
import './y.js'
```

```js
// y.js
console.log('foo')
```

If we build it with parcel `parcel serve --no-hmr index.html`, it includes/runs `y.js` two times.

Same time, if we use native ES modules:

```html
<script>
var parcelRequire // hello #3545 
</script>
<script type="module" src="./x.js"></script>
<script type="module" src="./x.js"></script>
```

The `y.js` is run once.

## 🤔 Expected Behavior

`y.js` should be run once.

## 😯 Current Behavior

Runs `y.js` multiple times.

## 💁 Possible Solution

Same as browserify - it handles that fine.
Should be code-splitting engaged or something, to avoid dupes in output.
I wouldn't bundle per-entry, instead I'd split out shared imports and reuse them.

## 🔦 Context

[spect](https://github.com/spectjs/spect/tree/atoms) enables aspect-oriented approach to web-sites, so that for example `material-design` components are initialized separately from main application logic, and separate from "decorative" logic (and separately from i18n, dataschema etc).

## 💻 Code Sample

For example, that spect app logic can be loaded each by own entry:

```html
<script src="./materialize.js"></script>
<script src="./app.js"></script>
<script src="./decorations.js"></script>
<script src="./i18n.js"></script>
<script src="./microdata.js"></script>
```

Since each that entry includes `spect` internally:

```js
// materialize.js
import { use } from 'spect'
import { MDLFab } from '@material/fab'

use('.mdl-fab', el => el.mdlComponent = new MDLFab(el))
```

```js
// app.js
import { on } from 'spect'

on('.app-form', 'submit', e => {
   // ...send form
})
....
```

That results in duplication of `spect` imports N times - per each entry.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 1.12.3
| Node             |  -
| npm/Yarn         | ∀
| Operating System | ∀

**Comment by mischnic** [comment]
```html
<script src="./x.js"></script>
<script src="./x.js"></script>
```
Personally, I would except everything to run twice here.

The CommonJS environment that Parcel emulates (and that ES6 imports are essentially transpiled to - doing `require("foo")` twice also executes that module only once) is only valid within that bundle.

**Comment by dy** [comment]
@mischnic 
For commonjs environment all bundlers provide code-splitting:
[browserify](https://github.com/browserify/factor-bundle), [rollup](https://rollupjs.org/guide/en/#code-splitting), [webpack](https://webpack.js.org/guides/code-splitting/). They have no problems sharing duplicate modules.
Also it's not the case of native ES modules, creates code duplication and problems of [global cache](https://github.com/ljharb/global-cache) / singletons.

What's the use of having isolated bundles? Does it solve any problem?

**Comment by devongovett** [comment]
Why are you importing the same script twice?

This is hard to do without wrapping every module in a function. When tree shaking/scope hoisting, this is exactly what we are trying to avoid.

I don't think this is possible.

**Comment by felipemullen** [comment]
I don't think this was justly closed. The code sample illustrated in the description does not include repetitions of the same script, but modules are claimed to have been duplicated in each.

That is the behavior I am seeing as well. I have a singleton being imported in a few files

```javascript
import { MyClass } from '../module'
```

And every file that uses the singleton ends up having its own separate copy of the module included in the output file. This does not appear to be standard behavior

**Comment by mischnic** [comment]
@felipemullen Could you please open a new issue with a reproduction?

**Comment by felipemullen** [comment]
@mischnic I've created a demonstration here: https://github.com/felipemullen/parcel-dupes

In the console output, you can see that modules `A` and `B` are included in the output for both `run.js` and `run2.js` which is not expected, and I think satisfies this original issue.

I wasn't sure if I should create a new issue yet because I feel like this behavior wouldn't have been overlooked. I must be doing something wrong?

Anyway, I can create a new issue if you think it is more appropriate

**Comment by mischnic** [comment]
@felipemullen Module hoisting with multiple scripts
```html
    <script src="run.js"></script>
    <script src="run2.js"></script>
```
isn't supported by Parcel, mainly because few people have requested that and putting `import "run.js"; import "run2.js";` in an `index.js` file works as expected in this case. (unless you have some reasons against this?)
Without es modules, we would even have to pollute the global namespace to access the modules in the first script (so `C` and `D` in your case).

**Comment by felipemullen** [comment]
Ah, I see. That's a pity. For context, I'll explain the scenario, maybe it will help others who come across this situation:

I am working on an enterprise level angular 1.6 project, and I have been slowly modernizing it where I can (The goal is to move away from angular in the long run)

The issue came up because multiple angular services (in separate files) make use of functionality that was extracted into a module, resulting in the complete duplication of the module in each service.

I understand now, thank you for the clarification. I will find a way around it 🐒 

Parcel is a great project, many thanks to everyone behind it!
